### ISSUE TITLE
3.0 release discussion

### ISSUE THREAD

**Issue by TomBZombie** [Level-2/Dice#127]
I'm going to put together a 3.0 release which has the following changes:

- Make better use of  PHP features

- replace `addRule` with `addRules` that takes an array in the format of

```
$dice->addRules(['Foo' => ['shared' => true], 'Bar' => [...]);
```

instead of

```
$dice->addRule('Foo', ['shared' =>true]);
$dice->addRule('Bar', [..]);
```

This would allow removal of the json loader and allow just: `$dice->addRules(json_decode(file_get_contents('foo.json')));`

- Fix the #125 by following the suggestions there.

- Also allow `Dice::Constant` to allow passing constants e.g. 
['\Dice::Constant` => 'PDO::EMULATE_PREPARES']

**Comment by solleer** [comment]
I can help with this if you would like.

**Comment by garrettw** [comment]
In the container I've been working on, the way I improved config loading was to move `addRule()` and `getRule()` out to a `Config` class, and just call them `add()` and `get()`. It also has a `load()` method which I'm using to pull in a config from a PHP file, but I'll probably make it work more like your `addRules()` idea.
In any case, my idea was to make a Config object and pass that into the Dice constructor – to uphold SRP.

**Comment by adjenks** [comment]
Do you plan on following psr-11?

**Comment by TomBZombie** [comment]
I don't want to break BC with existing projects, but there is already a wrapper here: https://github.com/TomBZombie/Dice-Interop/blob/master/Interop.php / https://packagist.org/packages/tombzombie/dice-interop 
```

**Comment by dnaber-de** [comment]
> I don't want to break BC with existing projects
But doesn't a raise of the major version ( from `2.*` to `3.*`) indicate a backward incompatible change?

**Comment by solleer** [comment]
Yes. In order to make certain improvements it must break BC.

**Comment by TomBZombie** [comment]
It does, but we'll provide a wrapper that allows backward-compatibility. E.g.  you'd just need to replace `new \Dice\Dice` with `new \Dice\Legacy(new \Dice\Dice())` or similar for it to work.

We could implement `ContainerInterface` during the 2.0-3.0 release but a wrapper is a neater solution for any project anyway, that way Dice remains dependency free and doesn't require people to download the ContainerInterface package unless they need it.

That said, I'll add the wrapper to packagist as I thought it was on there.

**Comment by dnaber-de** [comment]
> It does, but we'll provide a wrapper that allows backward-compatibility.
Why that? A typical workflow is to depend my project on `level-2/dice:^2.0` which allows me to fetch updates inside the 2.* branch but not from the new major version 3. I think more important than a wrapper for BC would be a transparent version strategy and consistent versioning numbers as well as [a CHANGELOG.md](http://keepachangelog.com/en/0.3.0/) and a better documentation of the API, one for each mayor version.

**Comment by garrettw** [comment]
I think I get what @dnaber-de is saying, and I would agree – what's the point in providing a wrapper for the legacy API if a new major version is generally understood to break BC anyway?

**Comment by solleer** [comment]
Also the wrapper would not able to be substituted in cases with type hinting

**Comment by TomBZombie** [comment]
But if you're using `ContainerInterface` you shouldn't be type hinting `Dice\Dice`, you'd typehint `ContainerInterface` anyway.

> I think I get what @dnaber-de is saying, and I would agree – what's the point in providing a wrapper for the legacy API if a new major version is generally understood to break BC anyway?

That's a fair point, we don't really need a wrapper but in the past people have wanted to move to the newer version without rewriting their config.

**Comment by dnaber-de** [comment]
> That's a fair point, we don't really need a wrapper but in the past people have wanted to move to the newer version without rewriting their config.

That's what [semantic versioning](http://semver.org/) is for. Composer version constraints following the idea of semantic versioning. One can frequently fetch minor updates and patches without need to worry about breaking changes. But avoiding breaking changes at all costs prevent the project from making progress and improvements.

**Comment by TomBZombie** [comment]
Indeed and you're right but during previous BC breaking changes, people have asked for a way to have the new features without needing to rewrite all their container rules. It can be nice for people to have their cake and eat it and a simple wrapper is the simplest way to achieve that.

I've also added to the list the ability to pass in global variables. Globals are bad, yes, but it's better to pass in $_POST where it's needed than reference the superglobal in arbitrary classes.

**Comment by TomBZombie** [comment]
I've implemented the changes above. The idea going forward is to encourage JSON configuration where possible. I've removed the JSON loader entirely as it now supports:

```
$dice->addRules(json_decode(file_get_contents('foo.json'));
```

There's no reason to wrap this in another class.

However, I've also removed the `addRule` function and replaced it with `addRules` that takes  key/value pair sin place of multiple calls with two arguments. As above, people probably won't want to rewrite all their rules so a simple wrapper is the best way of keeping BC:

```
//Either use composition or inheritance.. inheritance allows this class to work with \Dice\Dice type hints
class Dice2Compat extends \Dice\Dice {
	public function addRule($name, $value) {
		$this->addRules([$name => $value]);
	}
}
```

**Comment by jpirnat** [comment]
I understand the addition of `addRules`, but why remove `addRule`? That seems like a very needless BC break to me, especially when you say you don't want to break BC with existing projects.

**Comment by solleer** [comment]
Can there be the ability to access parts of globals like `$_GET['sort']` or something like that

**Comment by TomBZombie** [comment]
@jpirnat  true, I'll revert it, the -Dev branch isn't a final release it's open to change/discussion 

@solleer  it possibly could but it does open it up to scope creep, what if someone wants $_GET['sort'] ?? null? or $_GET['product']['id'], it would probably be better to use a request object like most of the frameworks do ( http://api.symfony.com/3.1/Symfony/Component/HttpFoundation/Request.html for symfony's) which would allow any kind of custom behaviour without needing to include it in Dice

**Comment by solleer** [comment]
That is what ill do then, I have a custom object I made for that purpose

**Comment by TomBZombie** [comment]
Since quite a few changes have been made to the 2.0 branch since I started on 3.0, I've created a new branch and implemented these changes.

I've also changed the constants values to make JSON rules a little neater. Instead of, as suggested above:

````json
{
	"Router": {
		"constructParams": {
			"\Dice::GLOBALS": "_SERVER"
		}
	}
}
```

I removed the slash prefix:

````json
{
	"Router": {
		"constructParams": [
			{"Dice::GLOBALS": "_SERVER"}
		]
	}
}
```

This goes for `INSTANCE` and `CONSTANT` as well. 

I'm also going to eventually drop XML configuration support. It will still be supported for now, but I'd rather encourage people to use JSON.

**Comment by solleer** [comment]
@TomBZombie When is 3.0 going to be released and which branch is the most up to date
I want to switch over but I want to be confident all patches are up to date

**Comment by TomBZombie** [comment]
3.0-Dev-Rebase is the latest branch, it contains all changes from 3.0 and any fixes that have been added to the 2.0 branch since.
